### DISCUSSION TITLE
Create a plugin in a different language than Javascript?

### DISCUSSION THREAD

**Discussion by norbjd** [serverless/serverless#10977] [Q&A] [answered]
Hello :wave: 

The [docs about plugin creation](https://www.serverless.com/framework/docs/guides/plugins/creating-plugins) states that:

> The simplest way to create a Serverless Framework plugin is to write a JavaScript file.

Is there a way to create a Serverless Framework plugin in a different language? If so, are you aware of existing plugins like this?

The plugin "base" could be in JS, but I'm wondering if we can use another language like Go for the plugin "logic". It could help people with only a little knowledge in JS.

Thanks! :pray:

**Comment by pgrzesik** [comment, marked as answer]
Hello @norbjd - in theory, that would be possible - you would have to write logic in some other language, but then call that logic from a JS "wrapper" that will actually integrate with the Framework and conform to the plugin API. Do you have any specific use case that could be addressed this way?

**Reply by norbjd** [reply]
Thanks for the quick answer :pray:
Actually, this is more curiosity but it could work in my case, though I have multiple questions. If, for example, my main code is written in a compiled language, like Go:

- how to pass data (arguments) between the Javascript wrapper and Go?
- how to package the plugin:
  - should I compile the Go logic code to a binary and call it in Javascript?
  - do people getting my plugin from npm should compile the Go, or should I ship the binary **directly** with the plugin?

These questions are not limited to Go, but to all compiled languages I think.

**Reply by pgrzesik** [reply]
Hey @norbjd

> how to pass data (arguments) between the Javascript wrapper and Go?

You would have to call the Go binary e.g. as a separate process or use something like https://github.com/node-ffi/node-ffi or https://github.com/cmake-js/fastcall to call a shared library build in a language of your choice. You would also have to distribute it e.g. as a part of the package.

> do people getting my plugin from npm should compile the Go, or should I ship the binary directly with the plugin?
That would be up to you I guess, but the less steps users have to take, the better usually. 

In general, I would avoid writing a plugin in a different language unless you have a really compelling use case.

**Reply by norbjd** [reply]
Thanks for the answer! You're right, I feel like writing a plugin in a different language can be counterproductive.
